File: app/src/test/java/com/example/rossinyamaral/bank/statementsScreen/StatementsPresenterUnitTest.java
package com.example.rossinyamaral.bank.statementsScreen;

import com.example.rossinyamaral.bank.model.StatementModel;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

@RunWith(RobolectricTestRunner.class)
public class StatementsPresenterUnitTest {

    @Test
    public void presentData_with_validInput_shouldCall_displayData() {
        //Given
        StatementsPresenter presenter = new StatementsPresenter();
        StatementsResponse response = new StatementsResponse();
        response.statements = new ArrayList<>();
        StatementModel statementModel = new StatementModel();
        statementModel.title = "test";
        statementModel.desc = "test";
        statementModel.date = "2000-01-01";
        statementModel.value = 10;
        response.statements.add(statementModel);

        StatementsActivityInputSpy activityInputSpy = new StatementsActivityInputSpy();
        presenter.output = new WeakReference<StatementsActivityInput>(activityInputSpy);

        //When
        presenter.presentStatementsData(response);

        //Then
        Assert.assertTrue("When the valid input is passed to Presenter " +
                        "Then displayData should be called",
                activityInputSpy.isDisplayDataCalled);
    }

    @Test
    public void presentData_with_inValidInput_shouldNotCall_displayData() {
        //Given
        StatementsPresenter presenter = new StatementsPresenter();

        StatementsActivityInputSpy activityInputSpy = new StatementsActivityInputSpy();
        presenter.output = new WeakReference<StatementsActivityInput>(activityInputSpy);

        //When
        presenter.presentStatementsData(null);

        //Then
        Assert.assertFalse("When the valid input is passed to LoginPresenter " +
                        "Then displayLoginMetaData should NOT be called",
                activityInputSpy.isDisplayDataCalled);
    }

    private class StatementsActivityInputSpy implements StatementsActivityInput {
        boolean isDisplayDataCalled = false;
        StatementsViewModel viewModelCopy;
        @Override
        public void displayStatementsData(StatementsViewModel viewModel) {
            isDisplayDataCalled = true;
            viewModelCopy = viewModel;
        }
    }
}

File: app/src/test/java/dev/vitorpacheco/solutis/bankapp/statementsScreen/StatementsPresenterUnitTest.kt
package dev.vitorpacheco.solutis.bankapp.statementsScreen

import org.junit.Assert.*
import org.junit.Test
import java.lang.ref.WeakReference

class StatementsPresenterUnitTest {

    @Test
    fun `presentStatementsData com valores válidos o método displayStatementsData deve ser chamado`() {
        val statementsPresenter = StatementsPresenter()
        val statementsResponse = StatementsResponse()

        val statementsPresenterOutputSpy = StatementsPresenterOutputSpy()
        statementsPresenter.output = WeakReference(statementsPresenterOutputSpy)

        statementsPresenter.presentStatementsData(statementsResponse)

        assertTrue(statementsPresenterOutputSpy.isdisplayLoginMetaDataCalled)
    }

    @Test
    fun `presentLoginData com valores inválidos o método displayLoginData não deve ser chamado`() {
        val statementsPresenter = StatementsPresenter()

        val statementsPresenterOutputSpy = StatementsPresenterOutputSpy()
        statementsPresenter.output = WeakReference(statementsPresenterOutputSpy)

        statementsPresenter.presentStatementsData(null)

        assertFalse(statementsPresenterOutputSpy.isdisplayLoginMetaDataCalled)
    }

    private class StatementsPresenterOutputSpy : StatementsActivityInput {

        var isdisplayLoginMetaDataCalled = false
        var loginViewModelCopy: StatementsViewModel? = null

        override fun displayStatementsData(viewModel: StatementsViewModel?) {
            isdisplayLoginMetaDataCalled = true
            loginViewModelCopy = viewModel
        }

    }

}